### PR TITLE
WIP: smaller image by almost 1gig

### PR DESCRIPTION
**Pull request by joshgoebel**
Reduces image size by almost 1gig.

- Removes bundle cache
- Removes Rails `.git` folder from using Rails git vs gems
- Remove yarn cache
- Removes Google Chrome download after it's installed
- Final build is on buster-slim (no buildpack)
- Bundled gems are copied over to final build in pre-built binary form
- `--disconnect_rate=5` polls every 200ms for disconnects, not every 10ms, lowers CPU usage for dev

### CODE REVIEW

File: docker/init.dev.sh [hunk 1]
```diff
@@ -1,13 +1,5 @@
-echo "Install bundler stuff"
-bundle update --full-index --conservative exercism_config
-
-echo "Setup Exercism Config"
-bundle exec setup_exercism_config
-
 echo "Create and migrate DBs"
-bundle exec bin/rails db:create
-bundle exec bin/rails db:migrate
+bundle exec bin/rails db:create db:migrate
```
Comment (iHiD) [inline docker/init.dev.sh]: Does this cause `migrate` to still run even if `create` exits with "db already exists"?

Comment (joshgoebel) [reply docker/init.dev.sh]: Good question... LOL.  I can revert this is need be.

**Comment by joshgoebel** [comment]
Though this gets into how the images are intended to be used.   I think this size win is pretty impressive (for production, deploys, etc) but if someone wanted to develop inside Docker long-term and treat it more as a little "linux computer with it's own disk", ie shell in, run bundle update, etc... then the 2.3gig image might be truly preferable as it's more "something you could live in" where-as this build is intended to be thrown away when it's time for the next build.

At least it's optimized so it should build the gems towards the end now so hopefully more of the build cache could be used in the future.

File: Dockerfile.dev [hunk 2]
```diff
@@ -1,27 +1,62 @@
...
 
 WORKDIR /usr/src/app
 
+RUN gem install -N bundler:2.1.4
+
+# copy over gems from build
 COPY Gemfile Gemfile.lock ./
-RUN bundle install
+COPY --from=builder /usr/local/bundle /usr/local/bundle
+RUN bundle list
```
Comment (iHiD) [inline Dockerfile.dev]: What's the purpose of this?

Comment (joshgoebel) [reply Dockerfile.dev]: The run?  It can be removed.  Testing.

**Comment by iHiD** [comment]
I've switched this to a Draft PR. You can set it for review when you're ready :)

File: docker/init.sql [hunk 3]
```diff
@@ -1,13 +1,12 @@
-CREATE USER 'exercism_v3' IDENTIFIED BY 'exercism_v3';
-
-CREATE DATABASE `exercism_v3_development`;
+CREATE DATABASE IF NOT EXISTS `exercism_v3_development`;
 ALTER DATABASE `exercism_v3_development` CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
-GRANT ALL PRIVILEGES ON `exercism_v3_development`.* TO 'exercism_v3';
+/* implicetly create the user with our first DB to avoid errors if they already exist */
```
Comment (iHiD) [inline docker/init.sql]:
```suggestion
/* Implicitly create the user with our first DB to avoid errors if they already exist */
```

**Comment by SleeplessByte** [comment]
(@joshgoebel this is _really_ cool -- i'll hide this to remove noise but wanted to say it anyway)

File: README.md [hunk 4]
```diff
@@ -30,12 +30,36 @@ To execute the Dockerfile, run the following with your AWS keys:
...
+
+```sh
+# run rubocop to lint the codebase
+./docker/lint
+
+# run rake test
+./docker/run-tests
+
+# run rake test:system
+./docker/run-systems-tests
```
Comment (iHiD) [inline README.md]:
```suggestion
./docker/run-system-tests
```

File: docker/lib/utils.sh [hunk 5]
```diff
@@ -0,0 +1,13 @@
+
+
```
Comment (iHiD) [inline docker/lib/utils.sh]:
```suggestion
```

File: Procfile.docker.dev [hunk 6]
```diff
@@ -1,4 +1,4 @@
 server: bundle exec bin/rails server -b '0.0.0.0'
 assets: WEBPACKER_DEV_SERVER_HOST=0.0.0.0 bundle exec bin/webpack-dev-server
-anycable: RAILS_ENV=development bundle exec anycable --server-command "anycable-go --host=0.0.0.0 --port 3334 --redis_url='redis://redis:6379'"
-js-tests: yarn test-watch
+anycable: RAILS_ENV=development bundle exec anycable --server-command "anycable-go --disconnect_rate=5 --host=0.0.0.0 --port 3334 --redis_url='redis://redis:6379'"
```
Comment (iHiD) [inline Procfile.docker.dev:3]: What does this do?

File: Procfile.docker.dev [hunk 7]
```diff
@@ -1,4 +1,4 @@
...
+#js-tests: yarn test-watch
```
Comment (iHiD) [inline Procfile.docker.dev]: Is this a mistake?

File: dev.Dockerfile [hunk 8]
```diff
@@ -1,27 +1,78 @@
...
-RUN curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add - && \
-    echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list && \
-    curl -sL https://deb.nodesource.com/setup_14.x | bash - && \
-    apt-get install -y cmake nodejs yarn tmux
+RUN set -ex; \
+    apt-get update; \
+    apt-get install -y cmake;
 
-RUN curl -L -o - https://github.com/DarthSim/overmind/releases/download/v2.2.0/overmind-v2.2.0-linux-amd64.gz | gunzip > /usr/local/bin/overmind && \	
+# this work we can do early since it's just copyign a binary to the slim build later
```
Comment (iHiD) [inline dev.Dockerfile]:
```suggestion
# We can do this work early and then copy a binary to the slim build later
```

File: docker/init.dev.sh [hunk 9]
```diff
@@ -1,13 +1,12 @@
...
-bundle exec setup_exercism_config
+while ! mysqladmin ping -h"$DB_HOST" --silent; do
+    echo "Waiting on MySQL..."
+    sleep 2
+done
 
 echo "Create and migrate DBs"
-bundle exec bin/rails db:create
+# init.sql creates initial MySQL USER and databases
+mysql -u root -ppassword -h "$DB_HOST" < docker/init.sql
```
Comment (iHiD) [inline docker/init.dev.sh]: This has moved to `docker/dev/`

File: docker/lib/utils.sh [hunk 10]
```diff
@@ -0,0 +1,11 @@
+handle_container_error() {
+result=$?
+if [ $result -ne 0 ]
+then
+    echo
+    echo "Hint: If you're told the 'container is not running' you'll need to start it first:"
+    echo
+    echo "    % docker-compose up -d website"
+    exit $result
+fi
```
Comment (iHiD) [inline docker/lib/utils.sh]: Should this all be indented a level?

File: docker/lint [hunk 11]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+# docker build -t exercism/website . -f ./dev.Dockerfile
```
Comment (iHiD) [inline docker/lint]: Should we remove this as it's irrelevant to this file?

File: docker/run-js-tests [hunk 12]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+# docker build -t exercism/website . -f ./dev.Dockerfile
```
Comment (iHiD) [inline docker/run-js-tests]: Should we remove this as it's irrelevant to this file?

File: docker/run-system-tests [hunk 13]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+# docker build -t exercism/website . -f ./dev.Dockerfile
```
Comment (iHiD) [inline docker/run-system-tests]: Should we remove this as it's irrelevant to this file?

File: docker/run-tests [hunk 14]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+# docker build -t exercism/website . -f ./dev.Dockerfile
```
Comment (iHiD) [inline docker/run-tests]: Should we remove this as it's irrelevant to this file?

File: docker/run-tests [hunk 15]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
```
Comment (iHiD) [inline docker/run-tests]: Move into `docker/dev/`

File: docker/run-system-tests [hunk 16]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
```
Comment (iHiD) [inline docker/run-system-tests]: Move into `docker/dev/`

File: docker/run-js-tests [hunk 17]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
Comment (iHiD) [inline docker/run-js-tests]: Move into `docker/dev/`

File: docker/lint [hunk 18]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
Comment (iHiD) [inline docker/lint]: Move into `docker/dev/`

File: docker/lib/utils.sh (hunk 10 above)
Comment (joshgoebel) [reply docker/lib/utils.sh]: Why not it's inside a block?

File: docker/lib/utils.sh [hunk 19]
```diff
@@ -0,0 +1,11 @@
+handle_container_error() {
```
Comment (iHiD) [inline docker/lib/utils.sh]: Move into `docker/dev/lib`

File: README.md [hunk 20]
```diff
@@ -30,10 +30,34 @@ To execute the Dockerfile, run the following with your AWS keys:
...
+
+Please check the repository README for additional help.
+
+#### Running tests in the scope of Docker Compose
+
+Assuming your docker-compose is already "up". All these are run "inside" the existing `website` container that needs to already have been started.
+
+```sh
+# run rubocop to lint the codebase
+./docker/lint
```
Comment (iHiD) [inline README.md]:
```suggestion
./docker/dev/lint
```

File: README.md [hunk 21]
```diff
@@ -30,10 +30,34 @@ To execute the Dockerfile, run the following with your AWS keys:
...
+
+# run rake test
+./docker/run-tests
```
Comment (iHiD) [inline README.md]:
```suggestion
./docker/dev/run-tests
```

File: README.md [hunk 22]
```diff
@@ -30,10 +30,34 @@ To execute the Dockerfile, run the following with your AWS keys:
...
+
+# run rake test:system
+./docker/run-system-tests
```
Comment (iHiD) [inline README.md]:
```suggestion
./docker/dev/run-system-tests
```

File: README.md [hunk 23]
```diff
@@ -30,10 +30,34 @@ To execute the Dockerfile, run the following with your AWS keys:
...
+
+# run yarn test
+./docker/run-js-tests
```
Comment (iHiD) [inline README.md]:
```suggestion
./docker/dev/run-js-tests
```

File: README.md (hunk 21 above)
Comment (joshgoebel) [reply README.md]: Isn't this sort of implied?  Are there production tests?

File: docker/lib/utils.sh (hunk 10 above)
Comment (iHiD) [reply docker/lib/utils.sh]: Is it not normal to indent from the `{` on L1?

File: Procfile.docker.dev (hunk 6 above)
Comment (joshgoebel) [reply Procfile.docker.dev:3]: It's mentioned in the main topic here:

> --disconnect_rate=5 polls every 200ms for disconnects, not every 10ms, lowers CPU usage for dev

File: docker/lib/utils.sh (hunk 10 above)
Comment (joshgoebel) [reply docker/lib/utils.sh]: Oh the whole fucntion, lol, I can do that also.

File: Procfile.docker.dev (hunk 6 above)
Comment (iHiD) [reply Procfile.docker.dev:3]: Sweet. Thanks.

**Comment by iHiD** [comment]
Amazing work, @joshgoebel. Thank you :)

**Pull request merged**